presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampStreamReader.java

```diff
@@ -238,14 +237,182 @@ private static int parseNanos(long serialized)
+                    break;
+                }
+                nextActive = inputPositions[activeIdx];
+                if (outputChannel != -1 && numResults > resultSizeBudget) {
+                    truncationRow = inputQualifyingSet.truncateAndReturnTruncationRow(activeIdx);
+                }
+                continue;
+            }
+            else {
+                // The row is not in the input qualifying set. Add to skip if non-null.
+                if (presentStream == null || present[i]) {
+                    toSkip++;
+                }
+            }
+        }
+        if (toSkip > 0) {
+            secondsStream.skip(toSkip);
+            nanosStream.skip(toSkip);
+        }
+        endScan(presentStream);
+    }
+
+    private void addResult(long value)
+    {
+        if (outputChannel == -1) {
+            return;
+        }
+
+        int position = numValues + numResults;
+        ensureValuesCapacity(position + 1, false);
+        values[position] = value;
+        if (valueIsNull != null) {
+            valueIsNull[position] = false;
+        }
+        numResults++;
+    }
+
+    @Override
+    protected void ensureValuesCapacity(int capacity, boolean includeNulls)
+    {
+        if (values.length < capacity) {
+            values = Arrays.copyOf(values, Math.max(capacity + 10, values.length * 2));
+            if (valueIsNull != null) {
+                valueIsNull = Arrays.copyOf(valueIsNull, values.length);
+            }
+        }
+        if (includeNulls && valueIsNull == null) {
+            valueIsNull = new boolean[values.length];
+        }
+    }
+
+    @Override
+    public int getFixedWidth()
+    {
+        return BIGINT.getFixedSize();
+    }
+
+    @Override
+    public Block getBlock(int numFirstRows, boolean mayReuse)
+    {
+        checkEnoughValues(numFirstRows);
+        if (mayReuse) {
+            return new LongArrayBlock(numFirstRows, valueIsNull == null ? Optional.empty() : Optional.of(valueIsNull), values);
+        }
+        if (numFirstRows < numValues || values.length > (int) (numFirstRows * 1.2)) {
+            return new LongArrayBlock(numFirstRows,
+                    valueIsNull == null ? Optional.empty() : Optional.of(Arrays.copyOf(valueIsNull, numFirstRows)),
+                    Arrays.copyOf(values, numFirstRows));
+        }
+        Block block = new LongArrayBlock(numFirstRows, valueIsNull == null ? Optional.empty() : Optional.of(valueIsNull), values);
+        values = null;
+        valueIsNull = null;
+        numValues = 0;
+        return block;
+    }
+
+    @Override
+    public void compactValues(int[] surviving, int base, int numSurviving)
+    {
+        if (outputChannel != -1) {
+            StreamReaders.compactArrays(surviving, base, numSurviving, values, valueIsNull);
+            numValues = base + numSurviving;
+        }
+        compactQualifyingSet(surviving, numSurviving);
+    }
+
+    @Override
+    public void erase(int numFirstRows)
```

Comment:
I think we can encapsulate this in a utility.

Reply:
@tdcmeehan This method changes member variable `numValues`, hence, I don't think it can be extracted into a static utility method. Technically, I could extract it into a protected method in `ColumnReader`: `protected void erase(int numFirstRows, Object values)`. My only reservation is that it will be used by scalar readers, but not by varchar, list, map, etc. readers, which might be confusing. What do you think?